### ISSUE TITLE
`no-unsafe-property-key`: false positive with type-guarded string values

### ISSUE THREAD

**Issue by rentalhost** [sindresorhus/eslint-plugin-unicorn#3509]
When trying to use a type-guarded value as an object property key, the rule triggers the error: *"Do not use an unsafe value as a property key. Use an explicit string or symbol key instead."*

While this makes perfect sense in many contexts, in my specific case, the `typeof` type guard guarantees that the value is a valid `string` at runtime and at the type level. The rule seems to ignore this type narrowing and flags it anyway.

Rule: `no-unsafe-property-key`

```ts
// If I remove `object` from the union, everything works fine and the error goes away.
// But notice that my value can be an `object` (or could be anything else) or a `string` in practice.
type Value = object | "test"; 

// Adding `as const` here doesn't change the outcome, I just added it to demonstrate that even this doesn't solve it.
const data = { test: 123 } as const;

export function example(input: Value) {
  // Note that here I am ensuring that `input` is safe-guarded to a `string`.
  // Inside the scope of the `if` block, `input` is narrowed down to `string`, or in this specific case, exclusively `"test"`.
  if (typeof input === "string") {
    // Here, `input` is still incorrectly flagged with the error:
    // `eslint: Do not use an unsafe value as a property key. Use an explicit string or symbol key instead.`
    return data[input];
  }

  return data.test;
}
